sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,171 @@ class GpuDynamicPartitionDataSingleWriter(
-          // concat the caches and this `batch`
-          val concat = withResource(subTables) { _ =>
-            withResource(batch) { _ =>
+          if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
+            // convert caches seq to tables and close caches seq
+            val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
+            // clear the caches
+            savedStatus.get.tableCaches.clear()
+            // concat the caches and this `batch`
+            val concat = withResource(subTables) { _ =>
               // append `batch` to sub tables
-              subTables += GpuColumnVector.from(batch)
+              subTables += table
+              withResource(Table.concatenate(subTables: _*)) { concat =>
+                GpuColumnVector.from(concat, outDataTypes)
+              }
             }
-            withResource(Table.concatenate(subTables: _*)) { concat =>
-              GpuColumnVector.from(concat, outDataTypes)
+            // write concat table
+            if (!needSplitBatch(
+              maxRecordsPerFile,
+              currentWriterStatus.recordsInFile,
+              concat.numRows))
+            {
+              statsTrackers.foreach(_.newBatch(concat))
+              currentWriterStatus.recordsInFile += concat.numRows()
+              // the `write` function will close `concat`
+              currentWriterStatus.outputWriter.write(
+                concat,
+                statsTrackers
+              )
+            } else {
+              withResource(concat){ _ =>
+                writeBatch(concat)
+              }
             }
-          }
-          // write concat table
-          if (!needSplitBatch(
+          } else if (!needSplitBatch(
             maxRecordsPerFile,
             currentWriterStatus.recordsInFile,
-            concat.numRows))
+            table.getRowCount().toInt))
           {
-            statsTrackers.foreach(_.newBatch(concat))
-            currentWriterStatus.recordsInFile += concat.numRows()
-            currentWriterStatus.outputWriter.write(concat, statsTrackers)
-          } else {
-            withResource(concat){concat =>  
-              writeBatch(concat)
+            val batch  = GpuColumnVector.from(table, outDataTypes)
+            closeOnExcept(batch) { _ =>
+              statsTrackers.foreach(_.newBatch(batch))
+              currentWriterStatus.recordsInFile += batch.numRows()
             }
-          }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            batch.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(batch))
-            currentWriterStatus.recordsInFile += batch.numRows()
             currentWriterStatus.outputWriter.write(batch, statsTrackers)
           } else {
-            withResource(batch){batch =>  
+            withResource(GpuColumnVector.from(table, outDataTypes)) { batch =>
```

Comment:
Not new with this PR, but this also has the problem of artificially holding onto the GPU memory despite the writer trying to free the batch's GPU memory during the write process.  This is probably best fixed in a separate PR, so I filed #6963.